### DISCUSSION TITLE
"Hints" mode to quickly select and act on arbitrary text patterns visible in the terminal

### DISCUSSION THREAD

**Discussion by rmacklin** [ghostty-org/ghostty#2394] [Feature Requests, Ideas] [answered]
_This feature has previously been mentioned in #189 - one of the comments there figured that issue would be a prerequisite, but Mitchell confirmed on [Discord](https://discord.com/channels/1005603569187160125/1005603569711452192/1262545040421421127) that it's not. Thus, I'm opening this separate issue for it (encouraged by Mitchell's desire to "hear more use cases/requests"). I personally don't have much need for the search scrollback feature, but I would have a hard time giving up kitty's hints feature._

For those who aren't familiar with kitty's hints feature, the corresponding documentation is at https://sw.kovidgoyal.net/kitty/kittens/hints/ (which also links to [its source code](https://github.com/kovidgoyal/kitty/tree/a38153c890229d72c02be4d29fd037bfdc28172c/kittens/hints)). From the intro:
> kitty has a _hints mode_ to select and act on arbitrary text snippets currently visible on the screen. For example, you can press `ctrl+shift+e` to choose any URL visible on the screen and then open it using your default web browser.
> 
> Similarly, you can press `ctrl+shift+p>f` to select anything that looks like a path or filename and then insert it into the terminal, very useful for picking files from the output of a git or ls command and adding them to the command line for the next command.

Those are just a couple examples, but through its flexible options you can go further. I don't leverage all the options, but I do use that second example practically on a daily basis, along with the similar `kitty_mod+p>h` shortcut to select a hash, which is very useful when working with git commands. I also use `kitty_mod+p>w` to select words, though not quite as often. In addition to the shortcuts for inserting the selection into the terminal at my current cursor position, I also have mappings to [copy](https://sw.kovidgoyal.net/kitty/kittens/hints/#cmdoption-kitty-kitten-hints-program:~:text=copy%20the%20match%20to%20the%20clipboard) each of these types of text to the clipboard (so I can paste them in e.g. a GitHub comment, Slack message, etc.).

---

FWIW, here are the relevant quotes from the comments in #189:

@ferdinandyb:
> But why I actually wanted to comment is related to generic search, but a bit more specific: sometimes you actually see what you need on screen, but want to quickly copy/paste/execute it without a mouse. Back when I was using kitty I used [hints](https://sw.kovidgoyal.net/kitty/kittens/hints/) a lot

@sluongng:
> I wanted to file a feature request for something like Kitty Hints

**Comment by rmacklin** [comment]
On [Discord](https://discord.com/channels/1005603569187160125/1140732081568223272/1268616921306234962) someone mentioned WezTerm also has a "quick select mode" and said:
> I tend to have to copy a lot of ips or hostnames to ssh into from output from scripts so if i could quickly copy those to use

I figured it would be useful to add a link the corresponding WezTerm docs in this issue - here they are: https://wezfurlong.org/wezterm/quickselect.html

I haven't used it myself, but it seems pretty similar to kitty's hints mode. One difference appears to be that the default action is to copy the selection to the clipboard while the action to insert it at the current terminal cursor (i.e. copy and paste the selection, as described in the docs) is achieved by typing the uppercase form of the prefix. In kitty, the [--program flag](https://sw.kovidgoyal.net/kitty/kittens/hints/#cmdoption-kitty-kitten-hints-program) can be configured in more general/flexible ways: in addition to inserting the selection, it can also copy to the clipboard or to a specified buffer, or run the default open program, or launch a specific program in a new kitty tab, window, overlay, etc. That said, I do think WezTerm has the two most common use cases covered.

Edit: It also seems like WezTerm matches all of the configured `quick_select_patterns`, whereas in kitty you usually filter to a specific kind of pattern like a hash or a file path, since you typically only want to insert a specific type of string at a particular time. But maybe this is possible in WezTerm despite my reading of their docs. Anyway, for Ghostty, I'd vote for making it possible to select a specific type. (If you know you want to select a hash and there are only a few hashes displayed but lots of other matches like file paths, highlighting only hash strings will result in fewer matches/less noise on the screen/shorter prefixes to type.)

**Reply by will** [reply]
Just for what it's worth, being able to add things `quick_select_patterns` was super helpful for me, because I work on a system that uses IDs in the form `/[a-z2-7]{25}[4aeimquy]/` and being able to copy them quickly is my favorite feature when using wezterm:


<img width="768" alt="image" src="https://github.com/user-attachments/assets/38b5107b-3482-4774-8a23-c5e82be0ff9d">

<img width="768" alt="image" src="https://github.com/user-attachments/assets/23ab48d0-2456-431b-81ac-e7fdcf5a816a">

**Reply by rmacklin** [reply]
You can do that in kitty, too. The distinction I noted is that it seemed to be that WezTern always matches against all of the patterns, so if you are only trying to select e.g. a hash, you would still get a bunch of matches for things that are not a hash. There didn't seem to be a way to bind one shortcut for selecting a hash, another for an IP address, another for a file path, etc.

So, I would hope that if this gets implemented in ghostty it can be the more flexible way that kitty offers.

**Reply by Grazfather** [reply]
This is indispensable for me as well. I have a few regexes for different types of selections, e.g. git hashes or file paths.

One additional thing that wezterm does it that it has a way to set the action to do on the match, so I can have a regex for URLs e.g. `"https?://[a-zA-Z0-9.]+[-a-zA-Z0-9()@:%_\\+.~#?&//=]*"`, and a key bind that highlights these and opens whichever I select in in my browser.

e.g. my config has:

```lua
  -- QuickSelect
  { key = "y", mods = "LEADER", action = act.QuickSelect },
  -- Open URL via Quick Select
  { key = "u", mods = "LEADER", action = act.QuickSelectArgs {
      label = "open url",
      patterns = { "https?://[a-zA-Z0-9.]+[-a-zA-Z0-9()@:%_\\+.~#?&//=]*" },
      action = wezterm.action_callback(function(win, pane)
        wezterm.open_with(win:get_selection_text_for_pane(pane))
      end),
    },
```

Also a config for which keys to prioritize would be nice, e.g. it could default to qwerty homerow keys, but for people who use another (such as myself) you could override it.

```lua
config.quick_select_alphabet = "arstneioqwfpzxcvluymdhgjbk"
```

**Reply by Grazfather** [reply]
I could see this done with a config like:
```
# Copy hashes
keybind = ctrl-x=hint:<regex>:copy
# Copy filenames
keybind = ctrl-y=hint:<otherregex>:copy
# Open urls
keybind = ctrl-z=hint:<urlregex>:open_browser
```

**Reply by SuperDuperDeou** [reply]
Ghostty already has links, it could leverage that.
```
keybind = ctrl-x=hint:regex:<regex>:copy
keybind = ctrl-y=hint:link:link-url:open_browser
```

**Reply by jwllzp** [reply]
it does but having a different `keybind` for every pattern can be overkill. Wezterm's implementation is simple. You turn on quick select mode and all matches are highlihgted along with a key to copy. You can hit the capitalized mode of a highlighted key to directly paste.

Keybinds do seem like an option for now, but the full feature is still missing.

**Reply by Grazfather** [reply]
I wouldn't want regexes for git hashes to be used for my keybind to open urls, though.

**Reply by dkasak** [reply]
Echoing others' thoughts about how useful this is. My own usage in kitty relies on its key arpeggio support, i.e. the ability to run the action based on a key *sequence* rather than simultaneously pressed keys. I then have an `Action x Pattern` family of bindings, such as:

```
# files/paths
map kitty_mod+o>f    kitten hints --program editor --type regex --regex "[a-zA-Z0-9./_~-]+"
map kitty_mod+p>f    kitten hints --program - --type regex --regex "[a-zA-Z0-9./_~-]+"
map kitty_mod+y>f    kitten hints --program @ --type regex --regex "[a-zA-Z0-9./_~-]+"

# hashes
map kitty_mod+p>h    kitten hints --program - --type hash
map kitty_mod+y>h    kitten hints --program @ --type hash
map kitty_mod+o>h    kitten hints --program plumb --type hash

# URLs
map kitty_mod+p>u    kitten hints --program - --type url
map kitty_mod+y>u    kitten hints --program @ --type url
map kitty_mod+o>u    kitten hints --program plumb --type url

# words
map kitty_mod+p>w    kitten hints --program - --type word
map kitty_mod+y>w    kitten hints --program @ --type word
map kitty_mod+o>w    kitten hints --program plumb --type word

# addresses (pointers)
map kitty_mod+p>a    kitten hints --program - --type regex --regex "0x[a-fA-F0-9]+"
map kitty_mod+y>a    kitten hints --program @ --type regex --regex "0x[a-fA-F0-9]+"
map kitty_mod+o>a    kitten hints --program plumb --type regex --regex "0x[a-fA-F0-9]+"

# MSCs
map kitty_mod+p>m    kitten hints --program - --type regex --regex "(?i:MSC)[0-9]{3,4}"
map kitty_mod+y>m    kitten hints --program @ --type regex --regex "(?i:MSC)[0-9]{3,4}"
map kitty_mod+o>m    kitten hints --program plumb --type regex --regex "(?i:MSC)[0-9]{3,4}"

# CVEs / GHSAs
map kitty_mod+p>c    kitten hints --program - --type regex --regex "(CVE-[0-9]{1,5}-[0-9]{1,5}|GHSA-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4})"
map kitty_mod+y>c    kitten hints --program @ --type regex --regex "(CVE-[0-9]{1,5}-[0-9]{1,5}|GHSA-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4})"
map kitty_mod+o>c    kitten hints --program plumb --type regex --regex "(CVE-[0-9]{1,5}-[0-9]{1,5}|GHSA-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4})"

# ELEMENTSEC / MATRIXSEC
map kitty_mod+p>s    kitten hints --program - --type regex --regex "((ELEMENT|MATRIX)SEC-[0-9]{4}-[0-9]{3,5})"
map kitty_mod+y>s    kitten hints --program @ --type regex --regex "((ELEMENT|MATRIX)SEC-[0-9]{4}-[0-9]{3,5})"
map kitty_mod+o>s    kitten hints --program plumb --type regex --regex "((ELEMENT|MATRIX)SEC-[0-9]{4}-[0-9]{3,5})"

# IPFS
map kitty_mod+p>i    kitten hints --program - --type regex --regex "(?:ipfs://)?Qm[a-zA-Z0-9]{44}"
map kitty_mod+y>i    kitten hints --program @ --type regex --regex "(?:ipfs://)?Qm[a-zA-Z0-9]{44}"
map kitty_mod+o>i    kitten hints --program plumb --type regex --regex "(?:ipfs://)?Qm[a-zA-Z0-9]{44}"

# RFCs
map kitty_mod+p>r    kitten hints --program - --type regex --regex "(?i:RFC)[ -]?[0-9]{3,4}"
map kitty_mod+y>r    kitten hints --program @ --type regex --regex "(?i:RFC)[ -]?[0-9]{3,4}"
map kitty_mod+o>r    kitten hints --program plumb --type regex --regex "(?i:RFC)[ -]?[0-9]{3,4}"
```

This gives me the ability to lead with `cmd+p`, `cmd+y` or  `cmd+o`, followed by a letter representing the pattern type, in order to print, copy or pipe (to an external program) the resulting match of any pattern type. Not everyone needs this kind of flexibility, but IMO it would be a shame to design it out of ghostty.

**Reply by gabeklavans** [reply]
yeah I use kitty's hints wayyyy too much to realistically give any terminal lacking a similar feature a shot. it's just so dang useful, and exactly the kind of feature I want my terminal emulator dealing with (copying what I see on the screen, no matter how it got there, in a quick keyboard-friendly way)

**Comment by KorigamiK** [comment]
The way foot and some other terms implement is to show lables over the links like this:
![Screenshot_2024-12-27-11-11-03_25601](https://github.com/user-attachments/assets/87ecb76e-7659-449e-9dbc-f2212a983719)

**Comment by howardjohn** [comment]
Adding my usage here since the rest seem to be about keyboard driven workflows. I use Alacritty currently with mouse hints. How it works is I can just hold shift, and any pattern I hover over becomes underlined. Clicking it does some action - copying it (IPs, UUIDs, etc) or opening it (files in IDE).

Config example:

```
[[hints.enabled]]
command = "/home/john/bin/goland-open"
post_processing = true
regex = '[-\w\d/\.]+?\.go:[0-9]*:[0-9]*: '

[[hints.enabled]]
action = "Copy"
post_processing = true
regex = '[-\w\d/\.]+?\.(go:[0-9]*|go|yaml|yml|md|sh|json|rs:[0-9]*|rs)'

[[hints.enabled]]
action = "Copy"
post_processing = false
regex = '([0-9a-f]{12,128})|([[:digit:]]{1,3}\.[[:digit:]]{1,3}\.[[:digit:]]{1,3}\.[[:digit:]]{1,3})'
```

**Reply by howardjohn** [reply]
I think https://github.com/ghostty-org/ghostty/discussions/4379 would actually resolve this

**Comment by rmacklin** [comment]
Just want to note this appears to also be a feature in 
iTerm2 per https://gitlab.com/gnachman/iterm2/-/issues/11410 and in Alacritty per
https://bitfehler.srht.site/posts/2023-09-28_git-commit-hash-hints-with-alacritty.html

I haven't used either of those terminal emulators so I'm not sure exactly how the feature compares to kitty's hints feature. But based on what I saw in those links it seems pretty similar. So a "hints mode" for selecting text patterns seems to be supported in four of the popular terminal editors that other folks are commonly coming to ghostty from (kitty, wezterm, alacritty, iterm2)!

(Someone above also mentioned foot terminal but from what I can see in https://codeberg.org/dnkl/foot#urls it seems to be limited to URLs; it doesn't seem like you can add your own regexes to select e.g. git SHAs or file paths. So not quite as flexible/powerful/useful as the others.)

**Comment by eliasnorrby** [comment]
Chiming in with another use case for kitty's hints. kitty allows you to [arbitrarily customize matching and actions tied to hints](https://sw.kovidgoyal.net/kitty/kittens/hints/#completely-customizing-the-matching-and-actions-of-the-kitten). I've got it set up to match GitHub PR/issue numbers (like `#2394`) so that they're effectively links I can select to view them in the browser. 

Customization is performed in a python script that's supplied when setting up a binding:

```
map ctrl+shift+o kitten hints --customize-processing open-pull-request.py
```

Because something like `#2394` lacks context about which repository the PR or issue relates to, the associated action is a [chain](https://github.com/eliasnorrby/dotfiles/blob/948bf238d21e439624103d6f3efd0f1a671cd367/terminal/kitty/open-pull-request.py) [of](https://github.com/eliasnorrby/dotfiles/blob/948bf238d21e439624103d6f3efd0f1a671cd367/terminal/kitty/get_pr_repository_url.sh) [scripts](https://github.com/eliasnorrby/dotfiles/blob/948bf238d21e439624103d6f3efd0f1a671cd367/shell/tmux/tmux_active_pane_directory.sh) that tries to determine that, in my case based on the path of the currently active tmux pane.

**Comment by jwllzp** [comment]
Soooo I realized that maybe I was trying to make ghostty something it wasn’t meant to be. I’ve been back to wezterm for a week now. Ghostty is meant to be the best drop in replacement for the standard terminal in your os, and niche configurations like this one belong elsewhere :)

**Reply by rmacklin** [reply]
I think you're mistaken about that; Mitchell has stated (on the changelog podcast and elsewhere) that his goals for ghostty are more ambitious than that. It's just going to take time to get there.

Not saying you shouldn't use wezterm for now (I'm still using kitty, mainly because it has hints - though also using ghostty, mostly via its quick terminal); just responding to the notion that this feature request isn't in line with ghostty's goals, since I don't think that's the case here.

**Reply by jwllzp** [reply]
Fair. The latest banner for ghostly is aligned with what you say. I was basing my comment from the original presentation.

**Comment by illfygli** [comment]
I also use the Kitty "hints" feature a lot, for disability/accessibility reasons.

I don't have a pointing device, all my tools are keyboard-first, and currently in Ghostty, I have to resort to "mouse keys", moving the otherwise absent cursor with the WASD keys, and clicking with F.

**Comment by pasteley** [comment]
I can see how this could also enable the implementation of native [tmux-fingers](https://github.com/morantron/tmux-fingers) — vim's `gw` like copy by two-character label

**Reply by 00-kat** [reply]
(Aside: `gw` in Vim is formatting of text; you're probably referring to some plugin.)

**Reply by Grazfather** [reply]
TIL tmux-fingers. This might allow me to actually use ghostty w/tmux until this is implemented (I was stuck on wezterm because I could not live without this feature).

**Reply by richard-sparq** [reply]
Oh that's brilliant - thanks for drawing attention to tmux-fingers!

**Reply by rmacklin** [reply]
Note that Mitchell has stated that he wants ghostty to obviate the reasons people run tmux in their local terminal (not on servers) since if you're always running tmux, there's a significant cost to that extra layer which kinda negates a lot of the benefits of running ghostty (in terms of performance as well as terminal features). And given the nature of a "hints"-like feature needing access to the scrollback, you'd need to be running tmux all the time for tmux-fingers to be useful.
Not saying you shouldn't use tmux-fingers, just pointing out for others who land here that there's a tradeoff in doing so.

**Reply by maximilianigl** [reply]
Agreed. I'll stick to Wezterm until ghostty supports it natively - to avoid this additional layer.

**Reply by Grazfather** [reply]
Same here. I still find wezterm + hints better than having to use tmux-fingers.

**Reply by rmacklin** [reply]
I'm in a similar boat, I still use kitty because of how often I use hints. But I do use ghostty's quick terminal for quick one-off things. Looking forward to the day I can fully switch.

**Reply by purajit** [reply]
Ah awesome - more `tmux-fingers` users! I had created https://github.com/ghostty-org/ghostty/discussions/7672 for this exact request and just closed that 
out because this one has a ton more traction. One of the biggest things that prevents me from not being 
in tmux permanently (session restoring and the customizable status bar being the others, though I assume 
I'll have to sacrifice the latter). I'm extraordinarily fond of the ergonomics of `tmux-fingers` compared to 
any of its alternatives.

With the amount of terminal features being surfaced modularly, I wonder if we'll reach a point where these 
types of extensions could be made by the community.

**Reply by Grazfather** [reply]
Unfortunately I never got tmux-fingers to work

**Reply by Aridor2006** [reply]
With some inspiration from https://github.com/ghostty-org/ghostty/discussions/3708#discussioncomment-11694361 and by using the standalone [tmux-thumbs](https://github.com/fcsonline/tmux-thumbs), this is a workaround I've been using on Mac:

```
keybind = "cmd+shift+h=write_screen_file:paste"
keybind = "cmd+shift+k=text:\x01cat \x05|thumbs -c -t /tmp/thumbs-selection.txt && pbcopy < /tmp/thumbs-selection.txt \n"
```

Same as in #3708, two step sequence: `cmd+shift+h` -> `cmd+shift+k`. Screen will be put into thumbs and the selected text will be copied to clipboard after.

**Reply by rmacklin** [reply]
You can do something similar with `kitten hints` as the intermediary.

This approach is still far less convenient/integrated compared to actually using hints in kitty, where it's one keybind to add the hint overlays and then as soon as you type the matching hint prefix, it inserts that value at your current cursor position.

So for example, you could've already typed `git diff ` and then triggered the keybind to select+insert a git SHA, then typed `..` and then triggered it again to select+insert another SHA - then you just type any remaining command line arguments and/or press enter. It's a very seamless experience, and your clipboard won't have been touched. And as mentioned [above](https://github.com/ghostty-org/ghostty/discussions/2394#discussioncomment-10859480), you can have separate keybinds for selecting a hash vs a file path vs a URL, etc., which is handy. (If you know you want to select a hash and there are only a few hashes displayed but lots of other matches like file paths, highlighting only hash strings will result in fewer matches/less noise on the screen/shorter prefixes to type.)

It would be awesome to have a similarly seamless experience in ghostty.

**Comment by con-f-use** [comment]
Hints are pretty much the feature I can't live without anymore and that's preventing me to switch away from Kitty.

**Comment by projectdp** [comment]
I wanted to add another use case example to the collection:

I tested and saw that Ghostty supports [OSC 8](https://github.com/Alhadis/OSC8-Adoption/) and I love that! This gets me part of the way there, but I like not touching the mouse if possible.

Recently I was recalling a feature that I loved in [rxvt-unicode ](https://software.schmorp.de/pkg/rxvt-unicode.html)that did this sort of hinting and was hoping I could get it set up with Ghostty. What you could do was set up an extension called [url-select](https://github.com/xyb3rt/urxvt-perls/tree/master/deprecated), similar to this: [Clickable URLs extension](https://wiki.archlinux.org/title/Rxvt-unicode#Clickable_URLs), but fully keyboard driven. I set it up with my own launcher and added some features to do different things with the matched text. At the time I had not seen other terminal emulators doing this sort of regex matching for hinting.

The main features I cared about:
1. Add arbitrary regexes to match various things in the terminal to act upon with different programs or actions: clipboard, browser, text editor, phone app, etc.
2. The ability to add a binding for a launcher or context action.
3. The hinting ability to check for matches, visibly clickable underlined links.

I added a whole bunch of regex matches that would help me with my daily work. I hope to get back to doing something like this on Ghostty. I recently implemented a workaround with my environment: Ghostty+Fish+fzf. I use fish functions to glue it all together. It kind of works but it doesn't really do hinting, instead it would regex over command history and command output, then I can fuzzy search through it and open URLs. It's not an ideal solution but I still prefer it over using the mouse.

If something like this does get added I think some improvements would be:
1. Ability to add compiled regexes.
2. Advanced hinting functionality:
- Example: With vimium you can toggle on a hints view which shows all URLs with shortcut keybinds. If implemented in Ghostty it could be paired with URLs or other arbitrary matches, and then be able to perform other actions contextually depending on what it matches.
- It would also be nice to be able to toggle on hints, select all items in a match type, for example if you wanted all IPs (IPv4 & IPv6) in the terminal, and copy them to the clipboard, rather than individually select one.
- Improved hint views with configurable options for different matches, different hint label patterns, and visibility options.

Generally here are some common match types to write regexes against, collected from my own use and this thread:
- URLs - to open in a browser.
- File paths / filenames - to insert into the shell prompt.
- Git commit hashes - to copy or use in git commands.
- Word matches - for quick selection or copying.
- IP addresses - for network or log analysis.
- UUIDs / GUIDs - for debugging or referencing identifiers.
- Ticket / issue numbers - to open in browsers or issue trackers.
- Email addresses - for copying or composing messages.

**Comment by pluiedev** [comment, marked as answer]
Hi everyone,

As with #4379, I'm splitting this thread into multiple discussions, since there's a few interlinked but separate feature requests at hand here. There's quite a bit of overlap between these two topics too, so hopefully we can come up with a more coherent set of features to support by splitting them into more manageable chunks.

- If you'd like to open URLs via keyboard input, please see #9568
- If you'd like to match arbitrary string patterns and map them into actions/URLs, please see #9545
- If there's any topic here that hasn't been spun off into its own thread, please reply below this post so I can create them (or create it yourself and link it here)

**Reply by dkasak** [reply]
Your #9545 doesn't accurately match what was requested by some (me included). What we are after is akin to what exists in kitty and includes:

- The ability to find and decorate ("hint") all matches for a given regex a user had specified in the config
- For this to trigger on a user-defined key binding
- For this to find all matches currently displayed on the screen, not just some matches
- The decorations should be short strings of numbers or letters, allowing the user to select a particular match by typing that short string
- Once selected, the match should be executed by either 1) being printed on the terminal, or 2) getting sent into the clipboard, or 3) getting piped to an arbitrary user-specified program

**Reply by pluiedev** [reply]
My point is that I don't think these should be grouped into one single function like Kitty, but rather a set of *related* functions that can be used independently to each other.

> The ability to find and decorate ("hint") all matches for a given regex a user had specified in the config
>
> For this to find all matches currently displayed on the screen, not just some matches

That would be #9545

> For this to trigger on a user-defined key binding
>
> The decorations should be short strings of numbers or letters, allowing the user to select a particular match by typing that short string

These count as #9568

> Once selected, the match should be executed by either 1) being printed on the terminal, or 2) getting sent into the clipboard, or 3) getting piped to an arbitrary user-specified program

Falls under #9546

**Reply by rmacklin** [reply]
> > Once selected, the match should be executed by either 1) being printed on the terminal, or 2) getting sent into the clipboard, or 3) getting piped to an arbitrary user-specified program
> 
> - Falls under #9546

This seems a little confusing since two of the main use cases for kitty hints are _not_ URLs: SHAs and file paths. e.g. We want to be able to select output that looks like a SHA when running git commands, or that looks like a file path when running any number of arbitrary commands.

**Reply by purajit** [reply]
Echoing all this, [tmux-fingers](https://github.com/Morantron/tmux-fingers) within tmux (the one I use right now, and from what I can tell, pretty similar to kitty's feature) is a good way for folks unfamiliar with this to test what the experience is. The right feature naming is probably closer to "mouseless copy/paste"

#9545 seems rather unrelated to this - this isn't about aliases; the hint is a way to quickly pick the target string visually, and has nothing to do with the underlying string. The hint is usually just 1 character, at most two. It's just an index.

#9546 _could_ cover the request here, if one of the possible "handlers" is a function that can print the string into the terminal buffer itself (similar to `tmux set-buffer`). Local terminal buffer operations is the most important thing in this request, not the ability to open/view URLs and files with custom programs - this is purely about handling the strings themselves within the terminal rather than running scripts or programs on them.

#9568 is good, but needs to be generalized. As rmacklin mentioned, I use this primarily for SHAs, UUIDs, and paths, and almost never for URLs. `tmux-fingers` even allows defining additional custom regexes (which I use to capture git branch names that use my branch name prefixes).

**Reply by projectdp** [reply]
> #9548 is good, but needs to be generalized. As rmacklin mentioned, I use this primarily for SHAs, UUIDs, and paths, and almost never for URLs. `tmux-fingers` even allows defining additional custom regexes (which I use to capture git branch names that use my branch name prefixes).

Did you mean #9568 ?

**Reply by projectdp** [reply]
> As with #4379, I'm splitting this thread into multiple discussions, since there's a few interlinked but separate feature requests at hand here. There's quite a bit of overlap between these two topics too, so hopefully we can come up with a more coherent set of features to support by splitting them into more manageable chunks.

After reading this thread and seeing your reply: 

https://github.com/ghostty-org/ghostty/discussions/4379#discussioncomment-14774633
> This is a design and scope question for me personally, and it seems like a lot of people have different views and needs on what this feature should entail. Should it just allow more URLs to be highlighted? Should it be able to parse arbitrary strings [and expand it to a link](https://github.com/ghostty-org/ghostty/discussions/4379#discussioncomment-14311099)? Should it be able to run [arbitrary scripts](https://github.com/ghostty-org/ghostty/discussions/4379#discussioncomment-14318243)? Etc. etc.
> 
> My personal standpoint is that arbitrary string matching and script running are out-of-scope for a terminal emulator (see also ["Ghostty is not a script runner"](https://github.com/ghostty-org/ghostty/discussions/2872#discussioncomment-12102667)), and people who rely on those sorts of expansions should devise some sort of tool in their text editors that emits [OSC 8](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda) sequences to explicitly tell the emulator which strings are supposed to be interpreted as hyperlinks. I'm open to allowing our URL regex to be more customizable, though.

It makes me think the people in this thread who care about the issue should be the ones to create more accurate and concise feature requests to address what we're looking for. My understanding is that this discussion was created to distill a variety of use cases into a comprehensive set of features to begin to add to Ghostty.

I think @dkasak @rmacklin and @purajit are on the right track, maybe we can collaborate on putting this together.

**Reply by Grazfather** [reply]
Seems that #9568 is the discussion for the key-driven aspect of this feature request.

**Reply by ackerleytng** [reply]
Not sure which thread I should be commenting on, but I'm looking forward to this feature :) 

What if we split it into these parts:

1. Building a list of "interesting" strings.
2. Selecting a string from the list.
3. Taking an action on it.

For (1), I think everyone has their own interesting strings, so this should pretty much be open to code. Shell out to some script, or some embedded language? I saw some other feature request to "transform" strings, like GH-1234 into the full github link. The code should be responsible for the transformation. 

For performance, we can run the code only on strings that are currently visible (I think the strings that are not visible aren't usually interacted with?). We could also make the range to scan configurable (like perhaps 50 lines above what's currently visible)

For (2), I'm thinking Vimium-style or tmux-fingers style, or maybe fzf style? I'd personally like vimium-style but I think this could also be configurable.

For (3), actions could be another keystroke for "open in browser" or yet another keystroke for "insert string in buffer". The full list of actions should also be configurable.

This flow is kind of modelled after [emacs' embark](https://github.com/oantolin/embark)

**Reply by Grazfather** [reply]
I think #13450 is probably what you want
